pkg/userop/testing/local_blockchain/node.go

```diff
@@ -0,0 +1,303 @@
+package local_blockchain
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"net/url"
+	"os"
+	"sync"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/shopspring/decimal"
+	"github.com/stretchr/testify/require"
+	"github.com/testcontainers/testcontainers-go"
+	"github.com/testcontainers/testcontainers-go/wait"
+
+	"github.com/layer-3/clearsync/pkg/abi/entry_point_v0_6_0"
+	"github.com/layer-3/clearsync/pkg/abi/kernel_ecdsa_validator_v2_2"
+	"github.com/layer-3/clearsync/pkg/abi/kernel_factory_v2_2"
+	"github.com/layer-3/clearsync/pkg/abi/kernel_v2_2"
+	"github.com/layer-3/clearsync/pkg/artifacts/session_key_validator"
+	"github.com/layer-3/clearsync/pkg/userop"
+)
+
+type EthNode struct {
+	Container    testcontainers.Container
+	Client       *ethclient.Client
+	LocalURL     url.URL
+	ContainerURL url.URL
+}
+
+var (
+	ethActiveNode  *EthNode
+	ethActiveUsers int64
+	ethMutex       sync.Mutex
+)
+
+func ethNodeClenaupFactory(ctx context.Context, t *testing.T) func() {
+	return func() {
+		ethMutex.Lock()
+		defer ethMutex.Unlock()
+
+		ethActiveUsers--
+		if ethActiveUsers <= 0 {
+			err := ethActiveNode.Container.Terminate(ctx)
+			require.NoError(t, err, "failed to terminate Go-Ethereum container")
+
+			ethActiveNode = nil
+		}
+	}
+}
+
+func NewEthNode(ctx context.Context, t *testing.T) *EthNode {
+	ethMutex.Lock()
+	defer ethMutex.Unlock()
+	slog.Info("starting Go-Ethereum node...")
+
+	if ethActiveNode != nil {
+		ethActiveUsers++
+		t.Cleanup(ethNodeClenaupFactory(ctx, t))
+
+		slog.Info("reusing existing Ethereum node")
+		return ethActiveNode
+	}
+
+	var (
+		err           error
+		gethContainer testcontainers.Container
+		ethClient     *ethclient.Client
+		rpcURL        *url.URL
+		containerURL  *url.URL
+	)
+
+	if u := os.Getenv("GETH_NODE_RPC_URL"); u != "" {
+		rpcURL, err = url.Parse(u)
+		require.NoError(t, err, "failed to parse local RPC URL")
+		containerURL = rpcURL
+	} else {
+		// TODO: use in-memory database instead of container volumes
+		gethContainer, err = testcontainers.GenericContainer(ctx, testcontainers.GenericContainerRequest{
+			ContainerRequest: testcontainers.ContainerRequest{
+				Image: "ethereum/client-go:v1.13.14",
```

Comment:
Maybe it is worth to extract the image or geth version to a constant and put in at the top of the file?

Reply:
I believe it's redundant, since the current implementation is quite straightforward in conveying that we're dealing with a container image, and abstracting that to a constant wouldn't necessarily enhance our understanding or readability of the code. The direct reference keeps the context clear and concise, especially for this singular usage.